R/CriteriaComparison.R

```diff
@@ -0,0 +1,362 @@
+      ) %>%
+      dplyr::left_join(ref.subset,
+        relationship = "many-to-many",
+        dplyr::join_by(
+          TADA.CharacteristicName, TADA.ResultMeasure.MeasureUnitCode,
+          TADA.ResultSampleFractionText, TADA.MethodSpeciationName
+        )
+      ) %>%
+      dplyr::rename_with(~ paste0("TADA.", group.id, "Name"), TADA.CharacteristicName) %>%
+      dplyr::rename_with(~ paste0("TADA.", group.id, ".ResultMeasureValue"), TADA.ResultMeasureValue) %>%
+      dplyr::rename_with(~ paste0("TADA.", group.id, ".MeasureUnitCode"), TADA.ResultMeasure.MeasureUnitCode) %>%
+      dplyr::rename_with(~ paste0("TADA.", group.id, "ActivityStartDateTime"), ActivityStartDateTime) %>%
+      dplyr::rename_with(~ paste0("TADA.", group.id, "ResultSampleFractionText"), TADA.ResultSampleFractionText) %>%
+      dplyr::rename_with(~ paste0("TADA.", group.id, "MethodSpeciationName"), TADA.MethodSpeciationName)
+
+
+
+    # identify dynamically named columns
+    pair_char_name <- paste0("TADA.", group.id, "Name")
+    pair_result_val <- paste0("TADA.", group.id, ".ResultMeasureValue")
+    pair_units <- paste0("TADA.", group.id, ".MeasureUnitCode")
+    pair_datetime <- paste0("TADA.", group.id, "ActivityStartDateTime")
+    pair_fraction <- paste0("TADA.", group.id, "ResultSampleFractionText")
+    pair_speciation <- paste0("TADA.", group.id, "MethodSpeciationName")
+
+    # NOTE - CAN REMOVE NCOUNT COLS WHEN DONE TESTING (HRM 8/20/24)
+
+    # pair by activity id
+    pair.activityid <- .data %>%
+      dplyr::filter(ActivityIdentifier %in% pair.subset$ActivityIdentifier) %>%
+      dplyr::left_join(pair.subset,
+        by = dplyr::join_by(ActivityIdentifier),
+        relationship = "many-to-many"
+      ) %>%
+      dplyr::group_by(ResultIdentifier) %>%
+      dplyr::slice_min(order_by = TADA.PairingGroup.Rank) %>%
+      dplyr::ungroup() %>%
+      dplyr::select(-TADA.PairingGroup.Rank) %>%
+      dplyr::group_by(ResultIdentifier) %>%
+      dplyr::slice_sample(n = 1) %>%
+      dplyr::select(
+        ResultIdentifier,
+        !!rlang::sym(pair_datetime),
+        !!rlang::sym(pair_result_val),
+        !!rlang::sym(pair_units),
+        !!rlang::sym(pair_fraction),
+        !!rlang::sym(pair_speciation)
+      )
+
+    # drop activity id from pair subset
+    pair.subset2 <- pair.subset %>%
+      dplyr::select(-ActivityIdentifier)
+
+    # pair by monitoring location and time
+    pair.ml.time <- .data %>%
+      dplyr::filter(
+        !ResultIdentifier %in% pair.activityid$ResultIdentifier,
+        !is.na(ActivityStartDateTime),
+        MonitoringLocationIdentifier %in% pair.subset$MonitoringLocationIdentifier
+      ) %>%
+      dplyr::left_join(pair.subset2,
+        relationship = "many-to-many",
+        by = dplyr::join_by(MonitoringLocationIdentifier)
+      ) %>%
+      dplyr::group_by(ResultIdentifier) %>%
+      # Figure out fastest time comparison method - needs to be absolute time comparison
```

Comment:
Is this part of the code in the code function looking at the closest ActivityStartDateTime for the paired parameters? What happens in cases of ties (for example: If the ActivityStartDateTime is 12/1/2010 16:57 and there is a ph ActivityStartDateTime 12/1/2010 16:47 and 12/1/2010 17:07?

![image](https://github.com/user-attachments/assets/d362985c-9d93-4028-806d-28c9f2d3a6f8)

Reply:
The "slice_sample(n=1)" at line 336 selects a random sample. This means that at the end of the pairing function, only one result will remain to pair, even if there was a "tie" between pairing group rank and time difference that would have left the original TADA result with more than one temp/pH/hardness result to be paired with.

In most of the examples, that I looked at, line 336 didn't come into play as selecting the lowest ranked pairing group and the smallest time difference yielded only one result for pairing. However, including the last slice_sample does account for situations where more than one option remains.